wagtail/contrib/wagtailapi/serializers.py

```diff
@@ -2,171 +2,268 @@
 
-class PageSerializer(WagtailSerializer):
-    def serialize_object_metadata(self, request, page, show_details=False):
-        data = super(PageSerializer, self).serialize_object_metadata(request, page, show_details=show_details)
+    These fields are a common way to link tags to objects in Wagtail. The API
+    serializes these as a list of strings taken from the name attribute of each
+    tag.
 
-        # Add type
-        data['type'] = page.specific_class._meta.app_label + '.' + page.specific_class.__name__
+    Example:
 
-        return data
+    "tags": ["bird", "wagtail"]
+    """
+    def to_representation(self, value):
+        return list(value.all().order_by('name').values_list('name', flat=True))
 
-    def serialize_object(self, request, page, fields=frozenset(), extra_data=(), all_fields=False, show_details=False):
-        # Add parent
-        if show_details:
-            parent = page.get_parent()
 
-            site_pages = pages_for_site(request.site)
-            if site_pages.filter(id=parent.id).exists():
-                parent_class = parent.specific_class
+class BaseSerializer(serializers.ModelSerializer):
+    # Add StreamField to serializer_field_mapping
+    serializer_field_mapping = serializers.ModelSerializer.serializer_field_mapping.copy()
+    serializer_field_mapping.update({
+        wagtailcore_fields.StreamField: StreamField,
+    })
+    serializer_related_field = RelatedField
 
-                extra_data += (
-                    ('parent', OrderedDict([
-                        ('id', parent.id),
-                        ('meta', OrderedDict([
-                             ('type', parent_class._meta.app_label + '.' + parent_class.__name__),
-                             ('detail_url', ObjectDetailURL(parent_class, parent.id)),
-                        ])),
-                    ])),
-                )
+    meta = MetaField()
 
-        return super(PageSerializer, self).serialize_object(request, page, fields=fields, extra_data=extra_data, all_fields=all_fields, show_details=show_details)
+    def build_property_field(self, field_name, model_class):
+        # TaggableManager is not a Django field so it gets treated as a property
+        field = getattr(model_class, field_name)
+        if isinstance(field, _TaggableManager):
+            return TagsField, {}
 
+        return super(BaseSerializer, self).build_property_field(field_name, model_class)
 
-class DocumentSerializer(WagtailSerializer):
-    def serialize_object_metadata(self, request, document, show_details=False):
-        data = super(DocumentSerializer, self).serialize_object_metadata(request, document, show_details=show_details)
 
-        # Download URL
-        if show_details:
-            data['download_url'] = URLPath(document.url)
+class PageSerializer(BaseSerializer):
+    meta = PageMetaField()
+    parent = PageParentField(read_only=True)
 
-        return data
+    def build_relational_field(self, field_name, relation_info):
```

Comment:
I'd *almost* be tempted to:

* push this into the base class (perhaps inside a conditional so its only used for page instances)
* drop the `meta = MetaField()` on the base class
* Instead declare the `meta = MetaField()` explicitly on the `ImageSerializer`

Then there's a nice kinda symmetry of all the behavior on the base class, with each of the three concrete cases explicitly listing their fields.

I'd also consider `create(): NotImplemented` and `update(): NotImplemented` on the base class, plus some visually obvious commenting to make the "here's the base" and "here's our three concrete cases" clear.